osgar/drivers/lora.py

```diff
@@ -0,0 +1,104 @@
+"""
+  LoRa - Long Range Radio
+
+  This module handles communication with LoRa devices and automatic re-transmission
+  of received messages so that all robots work as re-transmission stations.
+  The messages have limited size 40 bytes.
+
+  At the moment driver supports only transmission of "pose2d" messages so the control
+  center (and all neighbor robots) know about each other.
+
+  Current protocol uses numbers for DeviceID (1, 2, ... 5) for source identification
+  and '|' for data separation. The whole packet is terminated by '\n'. The DeviceID
+  is added in front. The message history is collected in reverse order.
+
+  Example:
+     b'2|5|4|hello' - original message "hello" was sent by device ID=4 and retransmitted
+                      by device ID=5 and retransmitted again by device ID=2
+"""
+from datetime import timedelta
+
+from osgar.node import Node
+from osgar.bus import BusShutdownException
+
+ALIVE_MESSAGE = b'alive'
+
+
+def parse_lora_packet(packet):
+    """
+       Parse block of binary data. The block starts with integer
+       numbers interlaced by '|'. List of transmitter IDs is returned
+       with the actual payload data. In the case of missing transmitted
+       history header None is returned.
+    """
+    s = packet.split(b'|')
+    try:
+        return [int(x) for x in s[:-1]], s[-1].strip()
+    except ValueError:
+        print('Invalid LoRa packet:', packet)
+        return None, packet
+
+
+def split_lora_buffer(buf):
+    if b'\n' not in buf:
+        return buf, b''
+    i = buf.index(b'\n')
+    return buf[i+1:], buf[:i+1]
+
+
+class LoRa(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        self.device_id = config.get('device_id')  # None for "autodetect"
+        self.min_transmit_dt = timedelta(seconds=10)  # TODO config?
+        self.buf = b''
+        self.raw = b''  # should be defined by Node
+        self.last_transmit = None
+
+    def send_data(self, data):
+        self.last_transmit = self.publish('raw', data + b'\n')
+        return self.last_transmit
+
+    def update(self):
+        channel = super().update()  # define self.time
+        if channel == 'raw':
+            print(self.time, "update", channel, self.raw)
+            self.buf, packet = split_lora_buffer(self.buf + self.raw)
+            if len(packet) > 0:
+                addr, data = parse_lora_packet(packet)
+                if addr is not None and self.device_id is not None and self.device_id not in addr:
+                    print('re-transmit')
+                    self.send_data(self.raw.strip())  # re-transmit data from other robots
```

Comment:
The original question I had here was "Why?" but github got confused. I figured given the `print()` above that is is trying to retransmit something. I didn't know why. Now that I have better understanding of what is going on - why is it sending `self.raw`? Who knows how many unparsed packets are there, right? Shouldn't it do something with `addr` and `data`?

Reply:
Hopefully fixed with `self.recent_packets` variable.